mplc/multi_partner_learning.py

```diff
@@ -654,3 +655,152 @@ def init_multi_partner_learning_from_scenario(scenario, is_save_data=True):
+        )
+
+        self.epsilon = epsilon
+        self.K = scenario.dataset.num_classes
+        self.history_theta = [[None for _ in self.partners_list] for _ in range(self.epoch_count)]
+        self.history_theta_ = [[None for _ in self.partners_list] for _ in range(self.epoch_count)]
+        self.theta = [self.init_flip_proba() for _ in self.partners_list]
+        self.theta_ = [None for _ in self.partners_list]
+
+    def init_flip_proba(self):
+        identity = np.identity(self.K)
+        return identity * (1 - self.epsilon) + (1 - identity) * (self.epsilon / (self.K - 1))
+
+    def fit(self):
+        start = timer()
+        logger.info(
+            f"## Training and evaluating model on partners with ids: {['#' + str(p.id) for p in self.partners_list]}")
+
+        logger.info("(LFlip) Very first minibatch, init new models for each partner")
+        partners_models = self.init_with_models()
+
+        self.epoch_index = 0
+        while self.epoch_index < self.epoch_count:
+            self.split_in_minibatches()
+            self.minibatch_index = 0
+            while self.minibatch_index < self.minibatch_count:
+                logger.info(f"(LFLip) Minibatch n°{self.minibatch_index} of epoch n°{self.epoch_index}, "
+                            f"init aggregated model for each partner with models from previous round")
+
+                # Evaluate and store accuracy of mini-batch start model
+                model_to_evaluate = partners_models[0]
+                model_evaluation_val_data = self.evaluate_model(model_to_evaluate, self.val_data)
+                self.score_matrix_collective_models[self.epoch_index,
+                                                    self.minibatch_index] = model_evaluation_val_data[1]
+
+                for partner_index, partner in enumerate(self.partners_list):
+                    # Reference the partner's model
+                    partner_model = partners_models[partner_index]
+                    x_batch = self.minibatched_x_train[partner_index][self.minibatch_index]
+                    y_batch = self.minibatched_y_train[partner_index][self.minibatch_index]
+
+                    predictions = partner_model.predict(x_batch)
+                    self.theta_[partner_index] = predictions  # Initialize the theta_
+
+                    for idx, y in enumerate(y_batch):
+                        self.theta_[partner_index][idx, :] *= self.theta[partner_index][:, np.argmax(y)]
+                    self.theta_[partner_index] = normalize(self.theta_[partner_index], axis=0, norm='l1')
+                    self.history_theta_[self.epoch_index][partner_index] = self.theta_[partner_index]
+
+                    self.theta[partner_index] = self.theta_[partner_index].T.dot(y_batch)
+                    self.theta[partner_index] = normalize(self.theta[partner_index], axis=1, norm='l1')
+                    logger.debug(f"(LFlip) theta : {self.theta[partner_index]}")
```

Comment:
If think we can remove this log, as we can access the theta with history

Reply:
As you wish but it is a debug log, so it is not display by default